### ISSUE TITLE
Data loading

### ISSUE THREAD

**Issue by aaawtest** [Alex-ray/v2-universal-js-hmr-ssr-react-redux#2]
Hey Alex, nice work!

How would one go about fetch data before rendering the page within this framework?

Thanks in advance!

**Comment by Alex-ray** [comment]
Hey @aaawtest , this is an on going  debate right now in the community; I haven't seen a great solution yet and I would have to think about this a bit more to find a long term solution.

In the meantime you can add a `fetchData` method for each root component and do something like this.

```javascript
let reqUrl = location.pathname + location.search;

      getReduxPromise().then(()=> {
        let reduxState = escape(JSON.stringify(store.getState()));
        let html = ReactDOMServer.renderToString(
          <Provider store={store}>
            { <RoutingContext {...renderProps}/> }
          </Provider>

        );
        res.render('index', { html, reduxState });
      });
      function getReduxPromise () {
        let { query, params } = renderProps;
        let comp = renderProps.components[renderProps.components.length - 1].WrappedComponent;
        let promise = comp.fetchData ?
          comp.fetchData({ query, params, store, history }) :
          Promise.resolve();

        return promise;
      }
    }
```

There is a similar boilerplate doing that here https://github.com/mz026/universal-redux-template/blob/master/app/server/middlewares/universalRenderer.js#L46

as well as a small tutorial on how to do that here https://www.codementor.io/reactjs/tutorial/redux-server-rendering-react-router-universal-web-app.

I'll definitely let you know if I find a better solution though!

Cheers,

Alex

**Comment by Alex-ray** [comment]
@aaawtest , im going to close this one out but feel free to reach out with any questions whatsoever!

Cheers, 

Alex

**Comment by esperancaJS** [comment]
Hey @Alex-ray where is the best file to include this logic ?

Thanks for the starter!

**Comment by Alex-ray** [comment]
Hey @PedroEsperanca , I think the best place for this is in `renderPage` function in the `ssr.js` file here https://github.com/Alex-ray/v2-universal-js-hmr-ssr-react-redux/blob/master/src/server/ssr.js#L32 .

Would love to here how it goes; still haven't found a better solution.... might be time to make a lil library out of it 🙂

**Comment by esperancaJS** [comment]
I ended up putting some of the logic there, however I went with an approach using `{ matchPath } from 'react-router';` to get route params in the server, avoiding to load a bunch of FE components in the server (like I see most starters do, including the one you recommended above). 

At the moment I have a new routes file with a simple json just for the server, but I might still simplify a bit. I'll share with you once I'm happy with it!
